src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -33,27 +34,54 @@ object RuleUtils {
    * @return indexes built for this plan
    */
   def getCandidateIndexes(indexManager: IndexManager, plan: LogicalPlan): Seq[IndexLogEntry] = {
-    // Map of a signature provider to a signature generated for the given plan.
-    val signatureMap = mutable.Map[String, String]()
-
-    def signatureValid(entry: IndexLogEntry): Boolean = {
-      val sourcePlanSignatures = entry.source.plan.properties.fingerprint.properties.signatures
-      assert(sourcePlanSignatures.length == 1)
-      val sourcePlanSignature = sourcePlanSignatures.head
-
-      if (!signatureMap.contains(sourcePlanSignature.provider)) {
-        val signature = LogicalPlanSignatureProvider
-          .create(sourcePlanSignature.provider)
-          .signature(plan)
-        signatureMap.put(sourcePlanSignature.provider, signature)
-      }
-      signatureMap(sourcePlanSignature.provider).equals(sourcePlanSignature.value)
-    }
+    getLogicalRelation(plan) match {
+      case Some(r) =>
+        // Map of a signature provider to a signature generated for the given plan.
+        val signatureMap: mutable.Map[String, String] = mutable.Map()
+
+        def signatureValid(entry: IndexLogEntry): Boolean = {
+          val sourcePlanSignatures =
+            entry.source.plan.properties.fingerprint.properties.signatures
+          assert(sourcePlanSignatures.length == 1)
+          val sourcePlanSignature = sourcePlanSignatures.head
+
+          if (!signatureMap.contains(sourcePlanSignature.provider)) {
+            val signature = LogicalPlanSignatureProvider
+              .create(sourcePlanSignature.provider)
+              .signature(r)
+              .get // It is safe to call Option.get as signature() is
+            // called on a LogicalRelation node which is
+            // valid for signature computation
```

Comment:
Hmm, should we just call `signature` with the given logical plan and not put it into the map if it returns None (instead of relying on `getLogicalRelation`)? Then this function will just return empty Seq, which translates to no candidate indexes, which makes sense.

Reply:
Using `getLogicalRelation` here is required for correctness as we need to extract the relevant subplan for the query plan to calculate the signature and compare it with covering indexes' signatures.
Here is an explanation on why we need to extract the logical relation. Plz let me know if it makes sense or we need to further discuss it
:
Assume we have a query plan as:
	QueryPlan (P):  Scan(R: a, b, c) -> Filter(a>5) -> Project(b)
and we have two indexes defined as:	
	Covering Index 1 (IX1): R(a, b)
	Index 2 (IX2): Scan(a, b, c) -> Filter(a>100) -> Project(b, c)	
As the LogicalRelation node is the same for all 3 plans, `FileBasedSignatureProvider` generates the same signature for all three of them.
`IndexSignatureProvider` (introduced by this change) includes node names in signature calculation and as a result it generate the same signature for P and IX2 and a different signature for IX1.

In FilterIndex and JoinIndex rules, a candidate index has to meet below two conditions:
- Index has to be defined on the same LogicalRelation as the one query uses,
- Index has to cover all records (i.e. index plan should not have a Filter, Join, Agg or any other node that removes or modifies original relation's records).
    
As a result, when we look for indexes in FilterIndex and JoinIndex rules we need to: "Extract a single logical relation node from original query plan and compare the signature of that subplan with the signature of covering indexes".
This way the signature comparison will satisfy both conditions mentioned above for these rules and only returns covering indexes that include all records of original relation.
In the above example, this procedure of extracting logical relation from query plan and comparing its signature with index signatures (which are calculated and stored using IndexSignatureProvider) results in only marking IX1 as a candidate index.

Reply:
Sounds like this would only apply to existing filter / join rule? In that case, those rules should call `getLogicalRelation` and call `getCandidateIndexes` with the result of `getLogicalRelation`. `getCandidateIndexes` doesn't seem like the right place to have a tight dependency on `getLogicalRelation`.

Reply:
That is correct and your suggestion makes sense. I changed the code accordingly.